File: webapp/src/services/cache.ts
/**
 * Cache API 客户端（sq-0009-p6）
 * 对应后端 api/routers/cache.py 的 7 端点
 */
import { useApi } from '@/composables/useApi'

export interface CoverageCell {
  latest: string | null
  status: 'fresh' | 'stale' | 'missing'
}

export interface CoverageResponse {
  symbols: string[]
  periods: string[]
  matrix: Record<string, Record<string, CoverageCell>>
  summary: {
    total_symbols: number
    total_cells: number
    fresh_cells: number
    stale_cells: number
    missing_cells: number
  }
}

export interface SyncLogItem {
  id: number
  symbol: string
  period: string
  sync_type: 'manual' | 'backfill' | 'scheduled'
  status: 'running' | 'success' | 'failed' | 'partial'
  start_at: string
  end_at: string | null
  rows_existing: number
  rows_new: number
  rows_total: number
  error_message: string | null
  trigger_source: string | null
  created_at: string
}

export interface SyncLogListResponse {
  total: number
  page: number
  page_size: number
  items: SyncLogItem[]
}

export interface ScheduleStateItem {
  task_name: string
  last_run_at: string | null
  next_run_at: string | null
  last_status: string | null
  run_count: number
  fail_count: number
  enabled: number
  updated_at: string | null
}

export const cacheApi = {
  /** 33 品种 × 5 周期覆盖率矩阵 */
  coverage:           () => useApi().get<CoverageResponse>('/cache/coverage'),

  /** 拉取记录分页 */
  syncLogs:           (params?: { symbol?: string; sync_type?: string; status?: string; page?: number; page_size?: number }) =>
                        useApi().get<SyncLogListResponse>('/cache/sync/logs', params),

  /** 单条详情 */
  syncLogDetail:      (logId: number) => useApi().get<SyncLogItem>(`/cache/sync/logs/${logId}`),

  /** 调度状态列表 */
  schedule:           () => useApi().get<{ items: ScheduleStateItem[] }>('/cache/schedule'),

  /** 启停调度任务 */
  toggleSchedule:     (taskName: string, enabled: boolean) =>
                        useApi().post<{ task_name: string; enabled: boolean; updated: boolean }>(
                          `/cache/schedule/${taskName}/toggle?enabled=${enabled}`
                        ),

  /** 回填历史 */
  backfill:           (symbol: string, period: string, years?: number, days?: number) => {
                        const params = new URLSearchParams({ symbol, period })
                        if (years != null) params.set('years', String(years))
                        if (days != null) params.set('days', String(days))
                        return useApi().post<any>(`/cache/backfill?${params.toString()}`)
                      },

  /** sq-0008 旧端点 */
  status:             () => useApi().get<any>('/cache/status'),
}

File: api/schemas/cache.py
"""Pydantic schemas for cache 域（sq-0009-p4）

6 端点对应 6+ schemas：
- GET /cache/coverage          → CacheCoverageResponse
- GET /cache/sync/logs         → SyncLogListResponse / SyncLogListItem
- GET /cache/sync/logs/{id}    → SyncLogDetail
- GET /cache/schedule          → ScheduleStateListResponse / ScheduleStateItem
- POST /cache/schedule/.../toggle → ScheduleToggleRequest / ScheduleToggleResponse
- POST /cache/backfill         → BackfillRequest / BackfillResponse
"""
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


# ============================================================
# /cache/coverage
# ============================================================

class CoverageCell(BaseModel):
    """38 品种 × 5 周期矩阵中一个单元"""
    latest: Optional[str] = None         # '2026-06-07' 或 '2026-06-07 17:00:00'
    status: str = Field(..., description="fresh / stale / missing")


class CacheCoverageResponse(BaseModel):
    """GET /cache/coverage 响应"""
    symbols: List[str] = Field(..., description="33 品种代码列表")
    periods: List[str] = Field(..., description="5 周期列表（daily/5min/15min/30min/60min）")
    matrix: Dict[str, Dict[str, CoverageCell]] = Field(
        ..., description="matrix[symbol][period] = CoverageCell"
    )
    summary: Dict[str, int] = Field(
        ..., description="total_symbols / total_cells / fresh_cells / stale_cells / missing_cells"
    )


# ============================================================
# /cache/sync/logs
# ============================================================

class SyncLogListItem(BaseModel):
    """拉取记录单条（列表项）"""
    id: int
    symbol: str
    period: str
    sync_type: str                       # manual / backfill / scheduled
    status: str                          # running / success / failed / partial
    start_at: str
    end_at: Optional[str] = None
    rows_existing: int = 0
    rows_new: int = 0
    rows_total: int = 0
    error_message: Optional[str] = None
    trigger_source: Optional[str] = None
    created_at: str


class SyncLogListResponse(BaseModel):
    """GET /cache/sync/logs 分页响应"""
    total: int
    page: int
    page_size: int
    items: List[SyncLogListItem]


class SyncLogDetail(SyncLogListItem):
    """GET /cache/sync/logs/{log_id} 详情（结构同列表项）"""
    pass


# ============================================================
# /cache/schedule
# ============================================================

class ScheduleStateItem(BaseModel):
    """调度任务状态"""
    task_name: str                       # daily_sync / minute_sync_5min
    last_run_at: Optional[str] = None
    next_run_at: Optional[str] = None    # P5 APScheduler 接入后写入
    last_status: Optional[str] = None    # success / failed
    run_count: int = 0
    fail_count: int = 0
    enabled: int                         # 0 / 1
    updated_at: Optional[str] = None


class ScheduleStateListResponse(BaseModel):
    """GET /cache/schedule 响应"""
    items: List[ScheduleStateItem]


class ScheduleToggleRequest(BaseModel):
    """POST /cache/schedule/{task_name}/toggle 请求体

    注：实际 task_name 在 URL 路径里，请求体只放 enabled
    """
    enabled: bool


class ScheduleToggleResponse(BaseModel):
    """POST /cache/schedule/{task_name}/toggle 响应"""
    task_name: str
    enabled: bool
    updated: bool                        # True if 任务存在并更新成功


# ============================================================
# /cache/backfill
# ============================================================

class BackfillRequest(BaseModel):
    """POST /cache/backfill 请求体"""
    symbol: str                          # 'AG' / 'AU' / ...
    period: str                          # 'daily' / '5min' / '15min' / '30min' / '60min'
    years: Optional[int] = None          # period='daily' 时用
    days: Optional[int] = None           # period='5min/15min/30min/60min' 时用
    start_date: Optional[str] = None     # sq-0009-round-4 commit 9: 用户选择范围
    end_date: Optional[str] = None       # sq-0009-round-4 commit 9: 用户选择范围


class BackfillResponse(BaseModel):
    """POST /cache/backfill 响应"""
    symbol: str
    period: str
    rows_new: int
    status: str                          # success / failed
    years: Optional[int] = None
    days: Optional[int] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
